feat: send logs in opentelemetry format

IronCore864 opened the pull request.

IronCore864 commented:
Some notes:

`serviceLog.Entry` has no "level", but for otelcol it's better to set severity number/severity text. Although optional, without, them, it's not extremely useful in the centralized logging system's UI because most tools can filter logs by levels:

<img width="1313" height="766" alt="Screenshot 2025-07-17 at 16 03 04" src="https://github.com/user-attachments/assets/232a921b-f20e-40be-9646-33853b29935c" />

Possible choices:

- Parse log text, if it contains "error", "info" at the beginning, use it as level; if it does not exist, defaults to info.
- Set all as info.
- Don't set (in this PR).

benhoyt commented on internals/overlord/logstate/opentelemetry/opentelemetry.go:
```diff
@@ -0,0 +1,405 @@
+
+	// Add service.name attribute.
+	keyValuePairs = append(keyValuePairs, KeyValue{
+		Key:   "service.name",
```
This is called `pebble_service` in the Loki labels. Would it be good to be consistent with that, or is `service.name` a standard OpenTelemetry thing?

IronCore864 replied:
`service.name` is a standard OpenTelemetry thing. For example:

```json
{
  "body": "hello world log message",
  "id": "0hLbG3uuXKargNtNbgvROReXpjK",
  "timestamp": "2025-07-21T09:31:10.501084928Z",
  "attributes": {
    "otelTraceSampled": true,
    "code.line.number": 8,
    "code.file.path": "/Users/tiexin/work/test/main.py",
    "code.function.name": "hello_world",
    "otelServiceName": "unknown_service",
    "otelSpanID": "d92b6a1ba1d1b6d9",
    "otelTraceID": "eb37f86f263ee2f74de147770b76dd28"
  },
  "resources": {
    "service.name": "unknown_service",
    "telemetry.auto.version": "0.56b0",
    "telemetry.sdk.language": "python",
    "telemetry.sdk.name": "opentelemetry",
    "telemetry.sdk.version": "1.35.0"
  },
  "scope": {},
  "severity_text": "WARN",
  "severity_number": 13,
  "scope_name": "root",
  "scope_version": "",
  "span_id": "d92b6a1ba1d1b6d9",
  "trace_flags": 1,
  "trace_id": "eb37f86f263ee2f74de147770b76dd28"
}
```

benhoyt commented on internals/overlord/logstate/opentelemetry/opentelemetry.go:
```diff
@@ -0,0 +1,405 @@
+
+	// Add service.name attribute.
+	keyValuePairs = append(keyValuePairs, KeyValue{
+		Key:   "service.name",
```
This is called `pebble_service` in the Loki labels. Would it be good to be consistent with that, or is `service.name` a standard OpenTelemetry thing?

benhoyt replied:
This is called `pebble_service` in the Loki labels. Would it be good to be consistent with that, or is `service.name` a standard OpenTelemetry thing?

benhoyt commented on internals/overlord/logstate/opentelemetry/opentelemetry_test.go:
```diff
@@ -0,0 +1,385 @@
+
+	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		c.Assert(r.Method, Equals, http.MethodPost)
+		c.Assert(r.Header.Get("Content-Type"), Equals, "application/json")
```
We should probably assert on the custom `User-Agent` too.

benhoyt commented on internals/overlord/logstate/opentelemetry/opentelemetry_test.go:
```diff
@@ -0,0 +1,385 @@
+	defer server.Close()
+
+	client := opentelemetry.NewClient(&plan.LogTarget{Location: server.URL})
+	client.SetLabels("svc1", map[string]string{})
+	client.SetLabels("svc2", map[string]string{})
+	client.SetLabels("svc3", map[string]string{})
+	client.SetLabels("svc4", map[string]string{})
```
I think if we're going to have a separate `TestLabels` we should drop these calls. But we could just call `SetLabels` with a couple of labels on (say) svc4 and combine the two tests. Most of it will be the same.

benhoyt commented on internals/overlord/logstate/opentelemetry/opentelemetry_test.go:
```diff
@@ -0,0 +1,385 @@
+	}
+
+	err := client.Flush(context.Background())
+	c.Assert(err, IsNil)
```
If the request goes through, we're asserting in the handler that it's the expected string, but we're not actually asserting that 1 request did go through. We should have a numRequests variable that's incremented in the handler and asserted on here.

benhoyt reviewed (changes requested):
Looks like a good start.

IronCore864 commented on internals/overlord/logstate/opentelemetry/opentelemetry.go:
```diff
@@ -0,0 +1,405 @@
+	// If this field is not set then resource info is unknown.
+	Resource Resource `json:"resource"`
+	// A list of ScopeLogs that originate from a resource.
+	ScopeLogs []*ScopeLogs `json:"scope_logs,omitempty"`
```
[Here](https://github.com/open-telemetry/opentelemetry-collector/blob/3c0fd3946f70a0b1fa97813c39dbc4d91d95afa6/pdata/internal/data/protogen/logs/v1/logs.pb.go#L229) it seems with protobuf, it's `json=scopeLogs` (camelCase) and with pure json, it's `json:"scope_logs"` (snake_case). It seems in the test it is different (same for other places like TimeUnixNano mentioned in another comment). As discussed, I also tried to figure it out using the official go SDK, but no matter how I configure it, it sends protobuf instead of JSON, so I could not figure it out. I think I will stick to the tests, meaning `camelCase`.

IronCore864 replied:
`[]*someStruct` v.s. `[]someStruct`: I tried to respect the original code, for example, in the original code it is `[]*ScopeLogs` and `[]KeyValue`, so I kept that. These structs won't be too large indeed, but I'm not sure if I should be consistent within our code, or consistent with the original code. Since for all the namings and json keys we stick with the original code, I think here we should also be consistent with the original code?

benhoyt replied:
For this one I think it should be internally consistent, so let's keep the naming consistent with the original, but use `[]slice` everywhere here (the fields are only a subset too).

IronCore864 commented on internals/overlord/logstate/opentelemetry/opentelemetry.go:
```diff
@@ -0,0 +1,405 @@
+			}
+			scope := Scope{
+				Name:    "pebble",
+				Version: cmd.Version,
```
user-agent is already set in `req.Header`.

scope_name and scope_version are set by otel SDK, so here I was trying to set the same. After more investigation, it seems the otel Python SDK would set scope_name to "root" and "scope_version" to empty by default, so I think it's OK not to set the version here. I will remove the version, only leaving the name for the scope.

IronCore864 commented on internals/overlord/logstate/opentelemetry/opentelemetry.go:
```diff
@@ -0,0 +1,405 @@
+}
+
+func NewClientWithOptions(target *plan.LogTarget, options *ClientOptions) *Client {
+	options = fillDefaultOptions(options)
```
fill/modify/return is preferred for two reasons:
1. `options.RequestTimeout` is also used in `httpClient`, not only in `Client.options`
2. It's the same logic as how the loki package does it.

benhoyt replied:
I don't mind the existence of `fillDefaultOptions`, but it seems quite bad practice to modify the caller's struct (it's also bad that the `loki` package does this -- we should change this).

So how about we make a copy instead and fill the copy? Something like this:

```go
opts := *options // make a copy of caller's options
fillDefaultOptions(&opts) // note: no need for this to return a value
...
```

Same effect, but we haven't modified the caller's struct.

benhoyt replied:
This still hasn't been done. Please do a similar copy-then-fill change here to what you did in `gatherer.go`, to avoid modifying the caller's struct.

IronCore864 commented:
Finished a round of refactoring according to review, a few comments left open so that it's easier to read the reasons why I didn't make those changes.

- [ ] TODO: manual test with otelcol/signoz to make sure the logs are still arriving at the target correctly.

benhoyt commented on docs/reference/layer-specification.md:
```diff
@@ -253,12 +253,18 @@ log-targets:
     # For Loki, this needs to be the fully-qualified URL of the push API,
     # including the API endpoint, e.g.
     #     http://<ip-address>:3100/loki/api/v1/push
+    # For OpenTelemetry, this needs to use the HTTP port (the default value is
```
Nit:
```suggestion
    # For OpenTelemetry, this needs to include the TCP port (normally
```

benhoyt commented on docs/reference/layer-specification.md:
```diff
@@ -253,12 +253,18 @@ log-targets:
     # including the API endpoint, e.g.
     #     http://<ip-address>:3100/loki/api/v1/push
+    # For OpenTelemetry, this needs to use the HTTP port (the default value is
+    # 4318) without the API endpoint, e.g.
```
```suggestion
    # 4318) without the API endpoint, for example:
```

benhoyt commented on docs/reference/log-forwarding.md:
```diff
@@ -21,8 +21,8 @@ log-targets:
-- `type`: The type of log target. The only supported type is `loki`.
-- `location`: The URL of the remote log target. For Loki, this needs to be the fully-qualified URL of the push API, including the API endpoint. Use the format `http://<ip-address>:3100/loki/api/v1/push`.
+- `type`: The type of log target. Supported types are `loki` and `opentelemetry`.
+- `location`: The URL of the remote log target. For Loki, this needs to be the fully-qualified URL of the push API, including the API endpoint; use the format `http://<ip-address>:3100/loki/api/v1/push`. For OpenTelemetry, use the HTTP port (default 4318) without the API endpoint, for example: `http://<ip-address>:4318`.
```
```suggestion
- `location`: The URL of the remote log target. For Loki, this needs to be the fully-qualified URL of the push API, including the API endpoint; use the format `http://<ip-address>:3100/loki/api/v1/push`. For OpenTelemetry, include the TCP port (normally 4318) without the API endpoint, for example: `http://<ip-address>:4318`.
```

benhoyt commented on internals/overlord/logstate/opentelemetry/opentelemetry_test.go:
```diff
@@ -83,117 +83,97 @@ func (*suite) TestRequest(c *C) {
 	{
 		"resource": {
-			"attributes": [{"key": "service.name", "value": {"stringValue": "svc1"}}]
+			"attributes": null
```
Why did this change to null? Shouldn't service.name always be there?

IronCore864 replied:
Because in the previous comment, you mentioned that since we already have a dedicated test that tests `SetLabels`, I removed it from here, and the default label `service.name` is set in `SetLabels`, same as the Loki package. In `gatherer`, this won't be a problem, since when the plan is changed, it always sets labels.

benhoyt commented on internals/overlord/logstate/opentelemetry/opentelemetry.go:
```diff
@@ -20,162 +20,140 @@ import (
-type KeyValue struct {
-	Key   string         `json:"key"`
-	Value AttributeValue `json:"value"`
+// keyValue is a key-value pair that stores	attributes.
```
Nit:
```suggestion
// keyValue is a key-value pair that stores attributes.
```

benhoyt commented on internals/overlord/logstate/opentelemetry/opentelemetry.go:
```diff
@@ -264,37 +253,34 @@ func (c *Client) Flush(ctx context.Context) error {
+	logs := make([]resourceLogs, 0, len(serviceNames))
 	for _, serviceName := range serviceNames {
 		batch := serviceBatches[serviceName]
 		if len(batch) > 0 {
```
Nit: it would avoid an extra nesting level to use an early-continue here: `if len(batch) == 0 { continue }`

benhoyt commented on internals/overlord/logstate/opentelemetry/opentelemetry.go:
```diff
@@ -264,37 +253,34 @@ func (c *Client) Flush(ctx context.Context) error {
-			}
-			scopeLogs := []*ScopeLogs{
+			scope := instrumentationScope{Name: "pebble"}
+			scopeLogs := []scopeLogs{
```
Nit: let's use cuddled braces here.

benhoyt commented on internals/overlord/logstate/opentelemetry/opentelemetry.go:
```diff
@@ -305,13 +291,13 @@ func (c *Client) Flush(ctx context.Context) error {
 	}
 
-	req, err := http.NewRequestWithContext(ctx, "POST", c.target.Location+"/v1/logs", bytes.NewBuffer(jsonData))
+	req, err := http.NewRequestWithContext(ctx, "POST", c.options.Location+"/v1/logs", bytes.NewBuffer(jsonData))
```
We should probably make this match the loki config and have the user include `/v1/logs` in the plan/config location. Slightly annoying, but gives them more flexibility if it's somehow at a different path, and matches the Loki "location" option.

IronCore864 replied:
I kindly disagree, because Loki and OpenTelemetry are two different things, and we should follow their conventions respectively, not matching them.

In Loki, the official documentation suggests that the URL includes the endpoint "/loki/api/v1/push", as seen in [here](https://grafana.com/docs/loki/latest/send-data/promtail/configuration/#clients) and [here](https://grafana.com/docs/alloy/latest/tutorials/send-logs-to-loki/). Whereas in OpenTelemetry, the convention is without, see [here](https://opentelemetry.io/docs/collector/configuration/).

The users would be confused to have different URLs between their Pebble config and other manual/local/automation. On the other hand, it's less likely to be confused when they switch from Loki to OpenTelemetry and have to update the config, because that is expected, since they switched to a new tool.

benhoyt commented on internals/overlord/logstate/opentelemetry/opentelemetry.go:
```diff
@@ -0,0 +1,381 @@
+		return nil, fmt.Errorf("error creating request: %v", err)
+	}
+	req.Header.Set("Content-Type", "application/json")
+	req.Header.Set("User-Agent", fmt.Sprintf("pebble/%s", cmd.Version))
```
Per earlier discussion, let's avoid the weird dependency from internals to cmd (and avoid hardcoding "pebble" here). Instead, let's add a `UserAgent string` field to the `ClientOptions` struct and have the caller set that to `cmd.ProgramName + "/" + cmd.Version`. An empty string could either mean "don't set User-Agent" or we could default to "pebble".

IronCore864 replied:
Got it, thanks, I did not fully get the previous comment on the user agent...

benhoyt reviewed (commented):
Thanks for the updates. Pretty close now I think. A few more minor comments and a couple of suggestions.

dimaqq commented:
Please take a look at https://opentelemetry.io/docs/specs/otlp/#json-protobuf-encoding if you haven't already, there are a few gotchas in the OTLP format:
- [ ] traceid/spanid (the only randomly generated id): hex strings without `0x`, typically padded with zeros to exact size by implementations.
- [ ] enum values: integers
- [ ] JSON object keys: `lowerCamelCase`
- [ ] large integers: decimal strings preferred, ints accepted
- [ ] Content-Type: application/json

Examples are here: https://github.com/open-telemetry/opentelemetry-proto/blob/v1.7.0/examples/README.md

IIRC there's not much difference between OTLP 1.5.0 and 1.7.0, but please check, as 1.7.0 is relatively new.

benhoyt commented on internals/overlord/logstate/opentelemetry/opentelemetry.go:
```diff
@@ -0,0 +1,381 @@
+	message := strings.TrimSuffix(entry.Message, "\n")
+
+	return logRecord{
+		TimeUnixNano: fmt.Sprintf("%d", entry.Time.UnixNano()),
```
It's a bit more efficient to use the more specific `strconv.FormatInt(entry.Time.UnixNano(), 10)` function that the generalized `fmt.Sprintf`.

benhoyt commented on internals/overlord/logstate/opentelemetry/opentelemetry.go:
```diff
@@ -0,0 +1,381 @@
+		resource := resource{
+			Attributes: resourceAttributes,
+		}
+		scope := instrumentationScope{Name: "pebble"}
```
Similar as with user agent -- let's add `ScopeName string` to `ClientOptions` and pass this in from above and `cmd.ProgramName`, rather than hardcode it here.

benhoyt commented on internals/overlord/logstate/opentelemetry/opentelemetry.go:
```diff
@@ -0,0 +1,381 @@
+func (c *Client) sendBatch(ctx context.Context, payload payload) (*http.Response, error) {
+	jsonData, err := json.Marshal(payload)
+	if err != nil {
+		return nil, fmt.Errorf("error marshaling log batch: %v", err)
```
Let's change to the canonical error format with "cannot", eg: `cannot marshal log batch: %v`. Similar elsewhere in this function.

IronCore864 commented:
On the `fillDefaultOptions` issue, I made an error and changed the wrong place. Reverted the unnecessary change and pushed a new commit to refactor according to the latest review. Now it's ready.

benhoyt reviewed (approved):
Cool, thank you! Let's get Harry's external eyes on it before merging. I'll ask him.

hpidcock commented on internals/overlord/logstate/opentelemetry/opentelemetry.go:
```diff
@@ -0,0 +1,383 @@
+	if err != nil {
+		return resp, fmt.Errorf("cannot send logs: %v", err)
+	}
+	defer resp.Body.Close()
```
Why are you closing the body here? `handleServerResponse` does it.

hpidcock commented on internals/overlord/logstate/opentelemetry/opentelemetry.go:
```diff
@@ -0,0 +1,383 @@
+	return c.handleServerResponse(resp)
+}
+
+func (c *Client) sendBatch(ctx context.Context, payload payload) (*http.Response, error) {
```
Don't return http response. Handle everything here.

IronCore864 replied:
I referred to the existing Loki package to follow the same logic, but after thinking about it, I agree that a better design is that `sendBatch` should call `handleServerResponse` internally and return only an error, because this would:

- eliminate the body close defer issue
- keep response handling encapsulated within this function
- simplify the API/signature

I will refactor accordingly, thanks for pointing this out :)
